File: 297-serialize-and-deserialize-binary-tree/serialize-and-deserialize-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string data= "";
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            data+="#";
            TreeNode* frnt= q.front();
            q.pop();
            if(!frnt) data+="N";
            else{
                data+= to_string(frnt->val);
                q.push(frnt->left);
                q.push(frnt->right);
            }
        }
        cout<<data<<"\n";
        return data;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if(data[1]=='N'){
            return NULL;
        }
        string temp="";
        int i =1, sz=(int)data.length();
        queue<TreeNode*> q;
        while(i<data.length() && data[i]!='#') temp+=data[i++];
        TreeNode *root= new TreeNode(stoi(temp));
        q.push(root);
        while(!q.empty()){
            TreeNode *frnt= q.front();
            q.pop();
            ++i; temp="";
            while(data[i]!='#') temp+=data[i++];
            if(temp!="N"){
                frnt->left= new TreeNode(stoi(temp));
                q.push(frnt->left);
            }
            else frnt->left=NULL;
            ++i; temp="";
            while(i<data.length() && data[i]!='#') temp+=data[i++];
            if(temp!="N"){
                frnt->right= new TreeNode(stoi(temp));
                q.push(frnt->right);
            }
            else frnt->right=NULL;
        }
        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));

File: 297-serialize-and-deserialize-binary-tree/serialize-and-deserialize-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if(root==null) return "N";
        StringBuilder sb= new StringBuilder();
        Queue<TreeNode> q= new LinkedList<>();
        q.add(root);
        TreeNode curr;
        while(!q.isEmpty()){
            curr= q.poll();
            if(curr!=null){
                sb.append(curr.val).append(",");
                q.add(curr.left);
                q.add(curr.right);
            }
            else sb.append("N,");
        }
        return sb.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        String[] vals= data.split(",");
        if(vals[0].equals("N")) return null;
        TreeNode root= new TreeNode(Integer.parseInt(vals[0]));
        int idx=1;
        Queue<TreeNode> q= new LinkedList<>();
        q.add(root);
        TreeNode node;
        while(!q.isEmpty()){
            node= q.poll();
            if(!vals[idx].equals("N")){
                node.left= new TreeNode(Integer.parseInt(vals[idx]));
                q.add(node.left);
            }
            idx++;
            if(!vals[idx].equals("N")){
                node.right= new TreeNode(Integer.parseInt(vals[idx]));
                q.add(node.right);
            }
            idx++;
        } 
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));